File: ruby/test/test_zero.rb
# frozen_string_literal: true

require "minitest/autorun"
require "base_human"

# Zero-config pair over the frozen baseh-medium-v1 profile: no profile
# object and no key, just to_code and from_code.
class TestZero < Minitest::Test
  def medium
    @medium ||= BaseHuman::Baseh.new(BaseHuman.baseh_medium_v1)
  end

  def assert_baseh_error(code, &block)
    error = assert_raises(BaseHuman::BasehError, &block)
    assert_equal code, error.code
  end

  def test_matches_the_frozen_medium_profile_exactly
    assert_equal medium.encode(id: 0), BaseHuman.to_code(0)
    assert_equal medium.encode(id: 123_456_789), BaseHuman.to_code(123_456_789)
    assert_equal "ZZZZZZV", BaseHuman.to_code(481_890_303)
    assert_equal "000000C", BaseHuman.to_code(0)
  end

  def test_to_code_accepts_integer_and_decimal_string
    assert_equal BaseHuman.to_code(123_456_789), BaseHuman.to_code("123456789")
  end

  def test_to_code_rejects_bad_input_with_argument_error
    assert_raises(ArgumentError) { BaseHuman.to_code("12x3") }
    assert_raises(ArgumentError) { BaseHuman.to_code("") }
    assert_raises(ArgumentError) { BaseHuman.to_code(nil) }
    assert_raises(ArgumentError) { BaseHuman.to_code(1.5) }
  end

  def test_to_code_raises_on_out_of_range_and_blocklisted_ids
    assert_baseh_error("OUT_OF_RANGE") { BaseHuman.to_code(481_890_304) }
    # 1131 is reserved by the Medium blocklist.
    assert_baseh_error("BLOCKED_CODE") { BaseHuman.to_code(1131) }
  end

  def test_from_code_returns_an_integer_and_round_trips
    id = BaseHuman.from_code(BaseHuman.to_code(123_456_789))
    assert_instance_of Integer, id
    assert_equal 123_456_789, id
  end

  def test_from_code_accepts_lowercase_aliases_and_any_whitespace
    code = BaseHuman.to_code(123_456_789)
    assert_equal 123_456_789, BaseHuman.from_code(code.downcase)
    spaced = "  #{code[0, 3]} #{code[3, 2]}\t#{code[5..]} "
    assert_equal 123_456_789, BaseHuman.from_code(spaced)
    # Typed aliases decode to canonical values.
    assert_equal 0, BaseHuman.from_code("OOOOOOC")
  end

  def test_from_code_raises_on_invalid_input_with_no_correction
    assert_baseh_error("INVALID_CHECKSUM") { BaseHuman.from_code("0000000") }
    assert_baseh_error("INVALID_CHARACTER") { BaseHuman.from_code("!!!!!!!") }
    # B is not canonical in Medium and is not an alias; no correction guesses it.
    assert_baseh_error("INVALID_CHARACTER") { BaseHuman.from_code("B00000C") }
    assert_baseh_error("INVALID_LENGTH") { BaseHuman.from_code("") }
  end
end

File: python/tests/test_zero.py
"""Unit tests for the zero-config pair (zero.py), mirroring js/test/zero.test.ts."""

import os
import sys
import unittest

sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "src"))

from baseh import (  # noqa: E402
    BLOCKED_CODE,
    INVALID_CHARACTER,
    INVALID_CHECKSUM,
    INVALID_LENGTH,
    OUT_OF_RANGE,
    Baseh,
    BasehError,
    baseh_medium_v1,
    from_code,
    to_code,
)

_MEDIUM = Baseh(baseh_medium_v1())


def _throws_code(fn, code):
    try:
        fn()
    except BasehError as err:
        if err.code == code:
            return
        raise AssertionError(f"expected code {code}, got {err.code}")
    raise AssertionError(f"expected BasehError({code}), no error raised")


class TestZeroConfig(unittest.TestCase):
    def test_matches_frozen_medium_profile(self):
        self.assertEqual(to_code(0), _MEDIUM.encode(0))
        self.assertEqual(to_code(123456789), _MEDIUM.encode(123456789))
        self.assertEqual(to_code(481890303), "ZZZZZZV")
        self.assertEqual(to_code(0), "000000C")

    def test_to_code_accepts_int_and_decimal_string(self):
        self.assertEqual(to_code(123456789), to_code("123456789"))
        with self.assertRaises((TypeError, ValueError)):
            to_code(-1)
        with self.assertRaises((TypeError, ValueError)):
            to_code("12x3")
        with self.assertRaises((TypeError, ValueError)):
            to_code("")
        with self.assertRaises((TypeError, ValueError)):
            to_code(1.5)
        with self.assertRaises((TypeError, ValueError)):
            to_code(None)

    def test_to_code_out_of_range_and_blocked(self):
        _throws_code(lambda: to_code(481890304), OUT_OF_RANGE)
        # 1131 is reserved by the Medium blocklist.
        _throws_code(lambda: to_code(1131), BLOCKED_CODE)

    def test_from_code_round_trip(self):
        id = from_code(to_code(123456789))
        self.assertIsInstance(id, int)
        self.assertEqual(id, 123456789)

    def test_from_code_accepts_lowercase_aliases_and_whitespace(self):
        c = to_code(123456789)
        self.assertEqual(from_code(c.lower()), 123456789)
        spaced = "  " + c[:3] + " " + c[3:5] + "\t" + c[5:] + " "
        self.assertEqual(from_code(spaced), 123456789)
        # Typed aliases decode to canonical values.
        self.assertEqual(from_code("OOOOOOC"), 0)
        # A lowercase alias form of a real code resolves to the same id.
        self.assertEqual(from_code("74uy c19"), from_code("74UYC19"))

    def test_from_code_invalid_input_no_correction(self):
        _throws_code(lambda: from_code("0000000"), INVALID_CHECKSUM)
        _throws_code(lambda: from_code("!!!!!!!"), INVALID_CHARACTER)
        # B is not canonical in Medium and is not an alias; no correction guesses it.
        _throws_code(lambda: from_code("B00000C"), INVALID_CHARACTER)
        _throws_code(lambda: from_code(""), INVALID_LENGTH)


if __name__ == "__main__":
    unittest.main()
